# 01 - Data Understanding
Load the cleaned datasets and understand their structure, distributions, and overlap.

In [1]:
import sys, os
sys.path.insert(0, os.path.abspath('..'))
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from config import PROCESSED_DATA_DIR
from src.util import load_dataframe

## 1.1 Load cleaned data

In [2]:
retention = load_dataframe(os.path.join(PROCESSED_DATA_DIR, 'retention_cleaned.csv'))
bob = load_dataframe(os.path.join(PROCESSED_DATA_DIR, 'bob_cleaned.csv'))

Loaded: retention_cleaned.csv - 37,144 rows, 37 cols
Loaded: bob_cleaned.csv - 197,870 rows, 26 cols


## 1.2 Retention - Profile

In [3]:
print(f'Shape: {retention.shape}')
print(f'Memory: {retention.memory_usage(deep=True).sum() / 1e6:.1f} MB')
print(f'\nData types:')
print(retention.dtypes.value_counts())
print(f'\nNull counts (top 10):')
nulls = retention.isnull().sum()
nulls = nulls[nulls > 0].sort_values(ascending=False)
print(nulls.head(10))

Shape: (37144, 37)
Memory: 60.6 MB

Data types:
object     22
float64     8
int64       7
Name: count, dtype: int64

Null counts (top 10):
Pull Type                  16704
Customer Account Number     1216
Customer Name               1216
Expected Pull Date           115
Agreement End Date            50
dtype: int64


In [4]:
retention.head()

,Case ID,Case Title,Country,Pull VAN,New VAN,VAN,Number of Contracts,Machines,Branch,Customer Account Number,...,Expected Pull Date,is_churned,CompanySize_numeric,is_zero_van,churn_reason_group,Risk_missing,Pull Type_missing,Customer Tier_missing,CompanySize_missing,Case Origin_missing
0,CAS-43717-G6L3B8,Invoice Issue,United Kingdom,0.0,0.0,5947.320,1.0,1.0,Bedford,UK02-CGBA021060-L,...,30-Sep-26,0,175.0,0,Other,1,0,0,0,0
1,CAS-43723-L3Z6Y3,Invoice Issue,United Kingdom,0.0,0.0,1611.648,1.0,1.0,Bedford,UK02-CGBA021060-L,...,29-Sep-26,0,175.0,0,Other,1,0,0,0,0
2,CAS-19468-S8J5V9,Overdue services on HCD,United Kingdom,0.0,0.0,1611.648,1.0,1.0,Bedford,UK02-CGBA021060-L,...,29-Sep-26,0,175.0,0,Service Issue,1,0,0,0,0
3,CAS-43721-Y0M5C3,Invoice Issue,United Kingdom,0.0,0.0,4030.488,1.0,1.0,Bedford,UK02-CGBA021060-L,...,29-Sep-25,0,175.0,0,Other,1,0,0,0,0
4,CAS-19467-N4K8L8,Overdue services on HCD,United Kingdom,0.0,0.0,4030.488,1.0,1.0,Bedford,UK02-CGBA021060-L,...,29-Sep-25,0,175.0,0,Service Issue,1,0,0,0,0


In [5]:
print('Column dtypes:')
print(retention.dtypes.to_string())

Column dtypes:
Case ID                       object
Case Title                    object
Country                       object
Pull VAN                     float64
New VAN                      float64
VAN                          float64
Number of Contracts          float64
Machines                     float64
Branch                        object
Customer Account Number       object
Customer Name                 object
Agreement End Date            object
Pull Type                     object
Case Type                     object
Risk                          object
Current Status                object
Resolution Status             object
Number Of Repair Cases       float64
Number of OverdueServices    float64
CompanySize                   object
Customer Tier                 object
Case Origin                   object
Case Creation Date            object
Resolved Time                 object
Registered Time               object
Resolved Date                 object
Registered Date        

In [6]:
# target distribution
print('Target (is_churned) distribution:')
print(retention['is_churned'].value_counts())
print(f'\nChurn rate: {retention["is_churned"].mean():.1%}')

Target (is_churned) distribution:
is_churned
0    22101
1    15043
Name: count, dtype: int64

Churn rate: 40.5%


## 1.3 BoB - Profile

In [7]:
print(f'Shape: {bob.shape}')
print(f'Memory: {bob.memory_usage(deep=True).sum() / 1e6:.1f} MB')
print(f'\nData types:')
print(bob.dtypes.value_counts())

Shape: (197870, 26)
Memory: 270.8 MB

Data types:
object     20
float64     6
Name: count, dtype: int64


In [8]:
bob.head()

,account_number,company_sizing,postal_code,branch,vat_number,agreement_number,agreement_start_date,agreement_end_date,renewal_type,agreement_type,...,bpg,msdyn_product_number,product_name,service_interval,unit_amount,billing_interval,billing_period,machine,machine_variant,chemistry
0,UK02-CGBA029544-L,>1000,PE1 5FQ,Elmswell,Unknown,GBC4019041-Machine Services-327311,2025-03-17,2026-03-16,Automatic Renewal,Scheduled Billing,...,17. Ultrasonic,ISK400EAA-001483-001,SK400 EAA + 400L of KLEEN7960S + Waste Rustpro...,1.0,9812.14,1.0,monthly,SK400,AA,KLEEN7960S
1,UK02-CGBA029544-L,>1000,PE1 5FQ,Elmswell,Unknown,GBC4021962-Machine Services-332280,2025-04-14,2028-04-13,Automatic Renewal,Scheduled Billing,...,17. Ultrasonic,ISK600EAA-001356-001,SK600 EAA + 600L of KLEEN7960S + Waste Rustpro...,1.0,8830.64,1.0,monthly,SK600,AA,KLEEN7960S
2,UK02-CGBA029544-L,>1000,PE1 5FQ,Elmswell,Unknown,GBC4021962-Machine Services-332282,2025-04-14,2028-04-13,Automatic Renewal,Scheduled Billing,...,17. Ultrasonic,ISK600EAA-001356-001,SK600 EAA + 600L of KLEEN7960S + Waste Rustpro...,1.0,8830.64,1.0,monthly,SK600,AA,KLEEN7960S
3,UK02-CGBA124601-L,20-49,BT71 7ET,Lisburn,GB714768220,GBC4000950-Auto waste,2022-06-24,2026-06-23,Automatic Renewal,Scheduled Billing,...,40. Oil Filters,K0000-040259-000,Waste Paint Sludge Collection - 205,0.0,7686.88,1.0,monthly,NaN,NaN,NaN
4,UK02-CGBA034502-L,50-99,BT41 3UF,Lisburn,Unknown,GBC3028394-Chemistry,2024-11-04,2025-11-03,Automatic Renewal,Scheduled Billing,...,30. Allied Products,C0000-000468-001,Delivery of Std Thinners - 205Ltrs (6),1.0,6933.91,1.0,monthly,NaN,NaN,NaN


In [9]:
# key categorical distributions
print('system_status:', bob['system_status'].value_counts().to_dict())
print('\nline_of_business:', bob['line_of_business'].value_counts().to_dict())
print('\nrenewal_type:', bob['renewal_type'].value_counts().to_dict())

system_status: {'Active': 197870}

line_of_business: {'Machine Services': 126683, 'Auto waste': 57649, 'Oil': 7522, 'Chemistry': 5815, 'Allied': 180, 'Kleenwaste': 21}

renewal_type: {'Automatic Renewal': 196606, 'No Automatic Renewal': 1264}


## 1.4 Dataset overlap

In [10]:
# check how many accounts overlap between the two datasets
ret_accts = set(retention['Customer Account Number'].dropna().unique())
bob_accts = set(bob['account_number'].dropna().unique())

common = ret_accts & bob_accts
only_ret = ret_accts - bob_accts
only_bob = bob_accts - ret_accts

print(f'Retention unique accounts: {len(ret_accts):,}')
print(f'BoB unique accounts: {len(bob_accts):,}')
print(f'Common: {len(common):,}')
print(f'Only in Retention: {len(only_ret):,}')
print(f'Only in BoB: {len(only_bob):,}')
print(f'Overlap rate: {len(common)/len(ret_accts)*100:.1f}%')

Retention unique accounts: 20,338
BoB unique accounts: 23,761
Common: 4,708
Only in Retention: 15,630
Only in BoB: 19,053
Overlap rate: 23.1%


## 1.5 Key Observations

- **Retention**: ~37K resolved cases with ~40% churn rate
- **BoB**: ~198K active records, 24K unique accounts
- **Overlap**: ~26% of retention accounts appear in BoB
- **CompanySize** had Excel date corruption (now fixed)
- **Customer Tier** had inconsistency (now fixed)
- **Customer Since** was 100% null (dropped)